lib/kdl/include/kdl/result_for_each.h

```diff
@@ -0,0 +1,111 @@
+/*
+ Copyright 2020 Kristian Duske
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#pragma once
+
+#include "kdl/meta_utils.h"
+#include "kdl/overload.h"
+#include "kdl/result.h"
+
+#include <iterator>
+#include <vector>
+
+namespace kdl {
+
+    /**
+     * Applies the given lambda to each element in the given range and returns the result.
+     *
+     * The given lambda must return a result type. If the lambda returns a void result type, then
```

Comment:
I had trouble following this. I'd rewrite it as the following:

```
Applies the given lambda to each element in the given range and returns the result.

The given lambda must return a result type

- If the lambda returns a void result type, returns result<void, error_type>.
  The void (success) case is returned if *all* invocations of the given lambda return void (success).

- If the given lambda returns a non void result type, returns result<std::vector<int>, error_type>.
  The vector is returned if *all* invocations of the given lambda return success values, and contains
  those success values.

If any invocation of the given lambda fails, processing stops and that failure result is returned.
```

This is clear enough that the examples aren't really needed imo.

Btw I don't like the term "value type". This is too ambiguous, should be "success type".

Reply:
Ok, I have replaced my doc with a slightly revised version of your proposal.